Add article about Inheritance in 0.6 (#77)
ethereum/solidity-blog: _posts/2020-06-12-solidity-0.6-inheritance.md

---
layout: post
title:  "Solidity 0.6.x features: inheritance"
date:   2020-06-12
author: Elena Gesheva

> ---

Reviewer: Also misses the category (I think this is the right notation):
```suggestion
category: Explainers
---
```

## Explicitely `virtual` and `override`
Functions are no longer virtual by default. This means a call to a non-virtual function will always execute that function regardless of any other contract in the inheritance hierarchy. This reduces abiguity that exists in version 0.5 where all functions are implicity virtual, allowing to be overriden further down the inheritance structure. This is especially dangerous in big inheritance graphs where this abiguity can lead to unintended behaviour and bugs.

>     pragma solidity ^0.5.17;

Reviewer 2: ```suggestion
```solidity
    pragma solidity ^0.5.17;
```

>     pragma solidity ^0.5.17;
>     contract A {
>         uint public x;
>         function setValue(uint _x) public {
>             x = _x;
>         }
>     }
> 
>     contract B {
>         uint public y;
>         function setValue(uint _y) public {
>             y = _y;
>         }
>     }
> 
>     contract C is A, B {
>     }

Reviewer 2: ```suggestion
```solidity
    pragma solidity ^0.5.17;
    contract A {
        uint public x;
        function setValue(uint _x) public {
            x = _x;
        }
    }

    contract B {
        uint public y;
        function setValue(uint _y) public {
            y = _y;
        }
    }

    contract C is A, B {
    }
```
```

>     }

Reviewer 2: ```suggestion
    }
```
```

> For example in contract `C` above calling `setValue` is calling the most derived implementation from contract `B`. However this is non-obvious from the implementation.

Reviewer 2: ```suggestion
For example in contract ``C`` above calling ``setValue`` is calling the most derived implementation from contract ``B``. However this is non-obvious from the implementation.
```

Author: Isn't this double back-tick for restructured text only? In markdown isn't it the same as a single back-tick?

> With 0.6 for the example above the compiler will now raise a type error: `Derived contract must override function "setValue". Two or more base classes define functions with the same name and parameter types`. This forces us to rething which implementation to use and explicitely override the conflicting base ones:

Reviewer 2: ```suggestion
With 0.6 for the example above the compiler will now raise a type error: `Derived contract must override function "setValue". Two or more base classes define functions with the same name and parameter types`. This forces us to rethink which implementation to use and explicitly override the conflicting base ones:
```

    pragma solidity ^0.6.10;
    contract A {
        uint public x;
        function setValue(uint _x) public virtual {
            x = _x;
        }
    }

    contract B {
        uint public y;
        function setValue(uint _y) public virtual {
            y = _y;
        }
    }

    contract C is A, B {
        function setValue(uint _x) public override(A,B) {
            A.setValue(_x);
        }
    }

You can only override functions if they are marked `virtual`. Additionally any overriding function must be marked `override`. If that again should be overridable, it has to also be marked `virtual`.

> Note: Interface functions are implicitely virtual so when implementing an interface you have to explicitely override its functions in the implementation. There is an ongoing discussion on this design [here](https://github.com/ethereum/solidity/issues/8281).

Reviewer 2: ```suggestion
Note: Interface functions are implicitly virtual so when implementing an interface you have to explicitely override its functions in the implementation. There is an ongoing discussion on this design [here](https://github.com/ethereum/solidity/issues/8281).
```

## Abstract contracts
With version 0.5 contract that doesn't implement all of its functions is implicitely made abstract by the compiler:
    
    pragma solidity ^0.5.17;
    contract X {
        function setValue(uint _x) public virtual;
    }

With 0.6 this differentiation has to be explicit with the compiler generating error `contract X should be made abstract` 
    
    pragma solidity ^0.6.10;
    abstract contract X {
        function setValue(uint _x) public virtual;
    }

## Interfaces can inherit

> This functionality is new to version 0.6 and allows interface inheritance. The resulting interface is a superset of all inherited interfaces' functions which the contract has to implement.

Reviewer 2: "a superset" sounds like it could also include other functions. What about something like "the combination of "?

    pragma solidity ^0.6.10;
    interface X {
        function setValue(uint _x) external;
    }

    interface Y is X {
        function getValue() external returns (uint);
    }

    contract Z is Y {
        uint x;
        function setValue(uint _x) external override { x = _x; }
        function getValue() external override returns (uint) { return x; }
    }

Note that if the contract does not implement all functions it has to be marked as `abstract`.

    pragma solidity ^0.6.10;
    abstract contract Z is Y {
        uint x;
        function setValue(uint _x) external override { x = _x; }
    }

> ## Public variables can override external functions

Reviewer 2: If this is not new, we can also just drop it.

Author: I've rewritten it to position as an improved feature with some anti patterns from 0.5. See if you like the new version.

---
layout: post
title:  "Solidity 0.6.x features: inheritance"
date:   2020-06-12
author: Elena Gesheva
category: Explainers
---

Similar to object-oriented programming in Solidity - a contract-oriented language - the inheritance and polymorphism features are as widely adopted and critical for the language evolution. There is hardly any Solidity developer who hasn't used these language features in their contracts to decouple logic and increase code reuse. With version 0.6 of the language the main improvements introduced are to make existing rules explicit in addition to introducing interface inheritance and disallowing the dangerous state variable shadowing. The compiler continues to use C3 Linearization, see [Solidity documentation](https://solidity.readthedocs.io/en/latest/contracts.html?#inheritance) for inheritance.

## Explicit `virtual` and `override`

> Functions are no longer virtual by default. This means that a call to a non-virtual function will always execute that function regardless of any other contract in the inheritance hierarchy. This reduces abiguity that exists in version 0.5 where all functions are implicity virtual, allowing to be overriden further down the inheritance structure. This is especially dangerous in big inheritance graphs where this abiguity can lead to unintended behaviour and bugs.

Reviewer 3: ```suggestion
Functions are no longer virtual by default. This means that a call to a non-virtual function will always execute that function regardless of any other contract in the inheritance hierarchy. This reduces ambiguity that exists in version 0.5 where all functions are implicitly virtual, allowing to be overridden further down the inheritance structure. This is especially dangerous in big inheritance graphs where this ambiguity can lead to unintended behaviour and bugs.
```

    }

    contract C is A, B {
    }
```

With version 0.6 for the example above the compiler raises a type error: `Derived contract must override function "setValue". Two or more base classes define functions with the same name and parameter types`. The above is an example of multiple inheritance where the same function is inherited from multiple base classes - `A` and `B`. In this case, it has to be overridden and the bases have to be listed in the override specifier. Important to note that the order in `override(A,B)` does not matter - specifically it does not change the behaviour of `super` - this is still dictated by the C3 linearization of the inheritance graph which is determined by the order in the `contract C is A, B { ... }` declaration.

```solidity
    pragma solidity ^0.6.10;
    contract A {
        uint public x;
        function setValue(uint _x) public virtual {
            x = _x;
        }
    }

    contract B {
        uint public y;
        function setValue(uint _y) public virtual {
            y = _y;
        }
    }

    contract C is A, B {
        function setValue(uint _x) public override(A,B) {
            A.setValue(_x);
        }
    }
```

> Note that you can only override functions if they are marked `virtual`. Additionally any overriding function must be marked `override`. If that again should be overridable, it has to also be marked `virtual`.

Reviewer 3: ```suggestion
Note that you can only override functions if they are marked `virtual`. Additionally, any overriding function must be marked `override`. If that again should be overridable, it has to also be marked `virtual`.
```

`interface` functions are implicitly `virtual` so when implementing an interface you have to explicitly override its functions in the implementation. There is an ongoing discussion on this design [here](https://github.com/ethereum/solidity/issues/8281).

> Worth noting that the keyword `super` works as it did before - calls the function one level higher up in the flattened inheritance hierarchy. Also unchanged is `super` is not allowed on `external` functions.

Reviewer 3: ```suggestion
It's worth noting that the keyword `super` works as it did before: It calls the function one level higher up in the flattened inheritance hierarchy. Also unchanged: `super` is still not allowed on `external` functions.
```

    interface X {
        function setValue(uint _x) external;
    }

    interface Y is X {
        function getValue() external returns (uint);
    }

    contract Z is Y {
        uint x;
        function setValue(uint _x) external override { x = _x; }
        function getValue() external override returns (uint) { return x; }
    }
```

Note that if the contract does not implement all functions it has to be marked as `abstract`.

```solidity
    pragma solidity ^0.6.10;
    abstract contract Z is Y {
        uint x;
        function setValue(uint _x) external override { x = _x; }
    }
```

## Abstract contracts
With version 0.5 a contract that doesn't implement all of its functions is implicitly made abstract by the compiler:

```solidity
    pragma solidity ^0.5.17;
    contract X {
        function setValue(uint _x) public virtual;
    }
```

With 0.6 this differentiation has to be explicit with the compiler generating error `contract X should be made abstract` otherwise

```solidity
    pragma solidity ^0.6.10;
    abstract contract X {
        function setValue(uint _x) public virtual;
    }
```

## Public variables safer override for external functions

> Although this feature existed before 0.6 it is made safer for public state variables to override external functions only if the parameter and return types of the function match the getter function of the variable. With version 0.5 it is possible these were allowed to differ as shown in the following example:

Reviewer 3: These two sentences sound overly complex to me, I would try to simplify them or split them up if possible?

Author: Yes, I managed to abbreviate it to 
>Although this feature existed before 0.6 it is now safer with the added checks that the getter function of the variable (generated by the compiler) matches the parameter and return type of the overriden external function. With version 0.5 it is possible these were allowed to differ as shown in the following example:

Does this sound better?

Reviewer 3: yes!

---
layout: post
published: true
title:  "Solidity 0.6.x features: inheritance"

> date:   2020-06-12

Reviewer: Should the date be updated?

Author: Sure. Done

Reviewer: Sorry, make sure to also rename the file to match the date
